syntax = "proto3";

service ManticoreUI {
    rpc Terminate(ManticoreInstance) returns (TerminateResponse) {}
    rpc Start(CLIArguments) returns (ManticoreInstance) {}
    rpc TargetAddress(AddressRequest) returns (TargetResponse) {}
}

message UUID {
    string hexstr = 1;
}

message ManticoreInstance {
    UUID id = 2;
}

message TerminateResponse { 

    enum TerminateStatus {
        SUCCESS = 0;
        ALREADY_STOPPED = 1;
        INSTANCE_NOT_FOUND = 2;
    }

    TerminateStatus status = 3;
}

message CLIArguments {
    string program_path = 4;
    repeated string binary_args = 5;
    string envp = 6;
    repeated string symbolic_files = 7;
    string concrete_start = 8;
    string stdin_size = 9;
    map<string, string> additional_mcore_args = 10;
}

message AddressRequest {
   
    enum TargetType {
        FIND = 0;
        AVOID = 1;
        CLEAR = 2;
    }
    
    ManticoreInstance mcore_instance = 11;
    uint64 address = 12;
    TargetType type = 13;
}

message TargetResponse {

    enum TargetStatus {
        SUCCESS = 0;
        ADDRESS_NOT_FOUND = 1;
        INSTANCE_NOT_FOUND = 2;
    }
    
    TargetStatus status = 14;
}
